File: getting-started-mocks/08-your-mock-is-lying/node/notifier/notifier.test.js
import assert from 'node:assert/strict'
import test from 'node:test'

import { CarrierContractError, CarrierUnavailableError } from './errors.js'
import { Notifier } from './notifier.js'
import { MemoryRecorder } from './recorder.js'

const dummySleep = async () => {}
const stubStatus = (status) => async () => status

function spySender() {
  const messages = []
  const send = async (_trackingNumber, message) => messages.push(message)
  send.messages = messages
  return send
}

test('notifies the customer only once', async () => {
  const send = spySender()
  const recorder = new MemoryRecorder()
  const notifier = new Notifier(stubStatus('delayed'), send, recorder, dummySleep)

  for (let i = 0; i < 3; i++) {
    await notifier.notify('TRACK-123')
  }

  assert.equal(send.messages.length, 1)
  assert.equal(recorder.count, 1)
})

test('does not record a delivered package', async () => {
  const send = spySender()
  const recorder = new MemoryRecorder()
  await new Notifier(stubStatus('delivered'), send, recorder, dummySleep).notify('TRACK-123')
  assert.deepEqual(send.messages, [])
  assert.equal(recorder.count, 0)
})

function mockSender(expectedMessage) {
  let calls = 0
  const send = async (_trackingNumber, message) => {
    calls++
    assert.equal(calls, 1, `sender called ${calls} times, expected exactly 1`)
    assert.equal(message, expectedMessage)
  }
  send.verify = () => assert.equal(calls, 1, `sender called ${calls} times, expected exactly 1`)
  return send
}

test('a strict mock owns its expectation', async () => {
  const send = mockSender('Package TRACK-123 is delayed')
  await new Notifier(stubStatus('delayed'), send, new MemoryRecorder(), dummySleep).notify('TRACK-123')
  send.verify()
})

test('surfaces a carrier failure without sending', async () => {
  const send = spySender()
  const broken = async () => {
    throw new CarrierUnavailableError('carrier returned 503')
  }
  const notifier = new Notifier(broken, send, new MemoryRecorder(), dummySleep)
  await assert.rejects(() => notifier.notify('TRACK-123'), CarrierUnavailableError)
  assert.deepEqual(send.messages, [])
})

test('does not retry a contract failure', async () => {
  let calls = 0
  const drifted = async () => {
    calls++
    throw new CarrierContractError('no status field')
  }
  const notifier = new Notifier(drifted, spySender(), new MemoryRecorder(), dummySleep)
  await assert.rejects(() => notifier.notify('TRACK-123'), CarrierContractError)
  assert.equal(calls, 1)
})

File: getting-started-mocks/06-they-arent-all-mocks/go/notifier/notifier_test.go
package notifier

import (
	"errors"
	"testing"
	"time"
)

func dummySleeper(time.Duration) {}

func stubStatus(status string) ShipmentStatus {
	return func(string) (string, error) { return status, nil }
}

type spySender struct{ messages []string }

func (s *spySender) Send(_ string, message string) error {
	s.messages = append(s.messages, message)
	return nil
}

func TestCustomerIsNotifiedOnlyOnce(t *testing.T) {
	sender := &spySender{}
	recorder := NewMemoryRecorder()
	n := New(stubStatus("delayed"), sender.Send, recorder, dummySleeper)

	for i := 0; i < 3; i++ {
		if err := n.Notify("TRACK-123"); err != nil {
			t.Fatal(err)
		}
	}

	if len(sender.messages) != 1 {
		t.Fatalf("sent %d messages, expected 1", len(sender.messages))
	}
	if recorder.Count() != 1 {
		t.Fatalf("recorded %d packages, expected 1", recorder.Count())
	}
}

func TestDeliveredPackageIsNotRecorded(t *testing.T) {
	sender := &spySender{}
	recorder := NewMemoryRecorder()
	n := New(stubStatus("delivered"), sender.Send, recorder, dummySleeper)

	if err := n.Notify("TRACK-123"); err != nil {
		t.Fatal(err)
	}
	if len(sender.messages) != 0 || recorder.Count() != 0 {
		t.Fatalf("unexpected notification: messages=%q records=%d", sender.messages, recorder.Count())
	}
}

type mockSender struct {
	t               *testing.T
	expectedMessage string
	calls           int
}

func (m *mockSender) Send(_ string, message string) error {
	m.t.Helper()
	m.calls++
	if m.calls > 1 {
		m.t.Fatalf("sender called %d times, expected exactly 1", m.calls)
	}
	if message != m.expectedMessage {
		m.t.Fatalf("sender got %q, expected %q", message, m.expectedMessage)
	}
	return nil
}

func (m *mockSender) verify() {
	m.t.Helper()
	if m.calls != 1 {
		m.t.Fatalf("sender called %d times, expected exactly 1", m.calls)
	}
}

func TestStrictMockOwnsItsExpectation(t *testing.T) {
	sender := &mockSender{t: t, expectedMessage: "Package TRACK-123 is delayed"}
	n := New(stubStatus("delayed"), sender.Send, NewMemoryRecorder(), dummySleeper)
	if err := n.Notify("TRACK-123"); err != nil {
		t.Fatal(err)
	}
	sender.verify()
}

func TestCarrierFailureSurfacesWithoutSending(t *testing.T) {
	sender := &spySender{}
	broken := func(string) (string, error) { return "", ErrCarrierUnavailable }
	n := New(broken, sender.Send, NewMemoryRecorder(), dummySleeper)
	if err := n.Notify("TRACK-123"); !errors.Is(err, ErrCarrierUnavailable) {
		t.Fatalf("expected carrier unavailable, got %v", err)
	}
	if len(sender.messages) != 0 {
		t.Fatalf("unexpected messages: %q", sender.messages)
	}
}
